src/main/java/net/dries007/tfc/common/blocks/plant/KelpTreeFlowerBlock.java

```diff
@@ -0,0 +1,355 @@
+    {
+        int i = rand.nextInt(5) + 1;
+        if (iterations == 0)
+        {
+            ++i;
+        }
+        for (int j = 0; j < i; ++j)
+        {
+            BlockPos blockpos = branchPos.above(j + 1);
+            if (!allNeighborsEmpty(worldIn, blockpos, null))
+            {
+                return;
+            }
+            setBodyBlockWithFluid(worldIn, blockpos, fluid);
+            setBodyBlockWithFluid(worldIn, blockpos.below(), fluid);
+        }
+
+        boolean willContinue = false;
+        if (iterations < 4)
+        {
+            int branchAttempts = rand.nextInt(4);
+            if (iterations == 0)
+            {
+                ++branchAttempts;
+            }
+
+            for (int k = 0; k < branchAttempts; ++k)
+            {
+                Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(rand);
+                BlockPos aboveRelativePos = branchPos.above(i).relative(direction);
+                if (Math.abs(aboveRelativePos.getX() - originalBranchPos.getX()) < maxHorizontalDistance && Math.abs(aboveRelativePos.getZ() - originalBranchPos.getZ()) < maxHorizontalDistance && isEmptyWaterBlock(worldIn, aboveRelativePos) && isEmptyWaterBlock(worldIn, aboveRelativePos.below()) && allNeighborsEmpty(worldIn, aboveRelativePos, direction.getOpposite()))
+                {
+                    willContinue = true;
+                    setBodyBlockWithFluid(worldIn, aboveRelativePos, fluid);
+                    setBodyBlockWithFluid(worldIn, aboveRelativePos.relative(direction.getOpposite()), fluid);
+                    growTreeRecursive(worldIn, aboveRelativePos, rand, originalBranchPos, maxHorizontalDistance, iterations + 1, fluid);
+                }
+            }
+        }
+        if (!willContinue)
+        {
+            worldIn.setBlock(branchPos.above(i), defaultBlockState().setValue(AGE, rand.nextInt(10) == 1 ? 3 : 5).setValue(getFluidProperty(), getFluidProperty().keyFor(fluid)), 2);
+        }
+    }
+
+    public static boolean isEmptyWaterBlock(IWorldReader worldIn, BlockPos pos)
+    {
+        return worldIn.isWaterAt(pos) && !(worldIn.getBlockState(pos).getBlock() instanceof IFluidLoggable);
+    }
+
+    private void setBodyBlockWithFluid(IWorld worldIn, BlockPos pos, Fluid fluid)
+    {
+        BlockState state = getBodyStateWithFluid(worldIn, pos, fluid);
+        worldIn.setBlock(pos, state, 2);
+    }
+
+    private BlockState getBodyStateWithFluid(IWorld worldIn, BlockPos pos, Fluid fluid)
+    {
+        KelpTreeBlock plant = (KelpTreeBlock) getBodyBlockAbstract().get();
+        return plant.getStateForPlacement(worldIn, pos).setValue(getFluidProperty(), getFluidProperty().keyFor(fluid));
+    }
+
+    @Override
+    public FluidProperty getFluidProperty()
+    {
+        return getFluidPropertyAbstract();
+    }
+
+    public abstract FluidProperty getFluidPropertyAbstract();
+
+    public abstract Supplier<? extends Block> getBodyBlockAbstract();
```

Comment:
Neither of these should need to exist, they can just be class level properties.

The hack with abstract properties in `PlantBlock` subclasses (and fluid properties) is that the `Block` constructor assembles block states, which needs references to all properties. This odd combination of `getFluidProperty` and `getFluidPropertyAbstract` is just not nessecary, and I see no use of `getBodyBlockAbstract` that would also predicate this.

Also calling methods `Abstract` is just a bad name overall - the method is never going to be abstract when it's implemented.